Build and infrastructure fixes for FreeBSD

tautschnig opened the pull request:
We rely on thin archives via `ar rcT` for libcprover-cpp nests archives, which FreeBSD's `ar` does not support. Use `llvm-ar` instead, which does have this capability. This enables a successful build.

All further changes are to enable successful test runs, with one caveat: goto-analyzer/loop-termination-eq/test-value-sets.desc yields a segmentation fault that I was unable to root cause thus far.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening commented on src/ansi-c/library/x86_assembler.c:
```diff
@@ -3,30 +3,30 @@
 extern int __CPROVER_rounding_mode;
 
-void __asm_fnstcw(unsigned short *dest)
+void __asm_fnstcw(void *dest)
```
Wozu denn diese Änderung?

tautschnig replied:
`remove_asmt::gcc_asm_function_call` produces function-type symbols with `void *` parameter types, and the above change avoids a linker warning. (Test execution on FreeBSD actually triggers this code path via `fesetround`.)

tautschnig replied:
Change removed, will PR separately.

tautschnig committed:
Build and documentation fixes for FreeBSD

We rely on thin archives via `ar rcT` for libcprover-cpp nests archives, which
FreeBSD's `ar` does not support. Use `llvm-ar` instead, which does have this
capability. This enables a successful build.

tautschnig committed:
Use /usr/bin/env in shebang lines

This improves portability as not all systems have bash, perl, or python
in /usr/bin (FreeBSD is one such example).

tautschnig committed:
C library check: exclude fenv on FreeBSD

FreeBSD has inline implementations of fe{set,get}round.

tautschnig committed:
C library: support inet_* redefinitions as found on FreeBSD

FreeBSD redefines various inet_* function names to __inet_*.

tautschnig commented:
> All further changes are to enable successful test runs, with one caveat: goto-analyzer/loop-termination-eq/test-value-sets.desc yields a segmentation fault that I was unable to root cause thus far.

#7956 fixes that one.

tautschnig committed:
cbmc-concurrency tests: FreeBSD would require sound pointer handling

This is the same problem as previously seen on macOS.

tautschnig committed:
FreeBSD supports hybrid binaries

It works for both FreeBSD kernel-only as well as full FreeBSD userland.

tautschnig committed:
libcprover-cpp regression test: don't duplicate object files

The object files are already part of libcprover-cpp, no need to add them
again. Doing so results in duplicate-definition errors on FreeBSD.

tautschnig committed:
C library: add FreeBSD's __srget, __swbuf

Implemented as over-approximations of the implementations found at
https://cgit.freebsd.org/src/tree/lib/libc/stdio/.

tautschnig commented:
@kroening @TGWDB I am hoping that this one is now uncontroversial. It would be great to get this merged in time for a version 6 release so that we can safely claim that version 6 also builds on FreeBSD. (And I would then try to sort out #7651 on top of this so that we can even claim OpenBSD/NetBSD support.)

TGWDB reviewed (approved):
I'd love and hate to have this tested in CI. Approving regardless of that.

tautschnig committed:
Run FreeBSD build and test in CI

Uses https://github.com/cross-platform-actions/action, and can be
extended to OpenBSD and NetBSD. QEMU on Linux runners appears to be
faster than nested virtualisation on macOS runnners (even when the
latter ought to use more hardware acceleration).

tautschnig commented:
> I'd love and hate to have this tested in CI.

Let's do it :-)

NlightNFotis commented:
Probably a naive question from my end, but what is the motivation to get CBMC officially supporting FreeBSD et al? Do people run verification tasks using these platforms or is there a need to support a downstream effort based on those?

tautschnig commented:
> Probably a naive question from my end, but what is the motivation to get CBMC officially supporting FreeBSD et al? Do people run verification tasks using these platforms or is there a need to support a downstream effort based on those?

We received https://github.com/model-checking/kani/issues/2710 on the Kani side. (And a more personal take: I firmly believe that writing portable code makes code better.)

tautschnig committed:
Pass CXX configuration to CaDiCaL configure

With Makefile-based builds, we rely on CaDiCaL's configure to produce an
appropriate build script.

tautschnig committed:
Do not hardcode GCC/g++ in shell scripts

BSD systems won't alias gcc to clang.

tautschnig committed:
Also build and test OpenBSD, NetBSD

Uses the same virtualisation set-up as FreeBSD checks.

NlightNFotis commented on .github/workflows/bsd.yaml:
```diff
@@ -60,3 +60,113 @@ jobs:
+
+  # This job takes approximately X to Y minutes
+  openbsd:
+    runs-on: macos-12
```
Is it possible to run these on linux machines rather than macOS ones?

I'm asking because I've noticed that macOS based runners are always slower (sometimes **much** slower) in acquiring a VM to run the action on.

tautschnig replied:
This is possible, but I was hoping that nested virtualisation (as available on macos, but not on Linux) would provide for better performance than QEMU. The performance, however, seems surprisingly poor, so perhaps going via QEMU isn't actually any slower. I'll experiment.

tautschnig committed:
Make memory_info portable across all BSD derivatives

We can always fall back to `getrusage` for minimal memory usage
information.

thomasspriggs commented:
I think the usage limits for github actions says that we are limited to 20 concurrent jobs, up to 5 of which may be macos jobs. It looks like we are over that limit. I guess this may mean jobs waiting for longer or be a reason to seek jobs to cull to the future. I am not arguing against testing building on BSD just noting this potential issue.

https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration

tautschnig committed:
Quote tags to avoid history expansion

Tags may contain an exclamation mark, which the shell will treat as a
request for history expansion.

tautschnig committed:
Remove inline assembler introduced by library functions

Library functions may introduce inline assembler. For example, FreeBSD's
fesetround is an inline function with inline assembler. Library
functions might introduce calls to fesetround, and thereby introduce
inline assembler. Thus, run a fixed point over remove_asm and library
linking.

tautschnig committed:
Use vmactions

This appears to have much better performance (FreeBSD job completes in
26 minutes instead of 4 hours). We can now also enable the same number
of steps in OpenBSD and NetBSD.

feliperodri commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1073,6 +1073,15 @@ void goto_instrument_parse_optionst::instrument_goto_program()
+    while(has_asm(goto_model))
+    {
+      remove_asm(goto_model);
+      link_to_library(
```
Should we include a test case to cover this new branch?

feliperodri commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -89,6 +89,15 @@ void cegis_verifiert::preprocess_goto_model()
   link_to_library(
     goto_model, log.get_message_handler(), cprover_c_library_factory);
+  // library functions may introduce inline assembler
+  while(has_asm(goto_model))
```
No need for regression tests here either?

tautschnig replied:
If we can get acceptable performance out of FreeBSD CI runs, we could exercise this. Still trying...

feliperodri reviewed (approved):
All changes in `goto-instrument` and `goto-synthesizer` LGTM module the lack of some regression tests (not a blocker).

tautschnig committed:
Support OpenBSD's __assert2 and NetBSD's __assert13

This has the same usage as __assert_func in newlib. See
https://github.com/openbsd/src/blob/master/include/assert.h.

tautschnig committed:
NetBSD: unit tests fail

This needs to be debugged eventually.
